sympy/solvers/solveset.py

```diff
@@ -1754,6 +1757,50 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _is_lambert(f, symbol):
+    r"""
+    Return True if the equation is of Lambert type, else False.
+    Equations containing `Pow`, `log` or `exp` terms are currently
+    treated as Lambert types.
+    """
+    return any(isinstance(arg, (Pow, exp, log)) for arg in _term_factors(f))
+
+
+def _compute_lambert_solutions(lhs, rhs, symbol):
+    """
+    Computes the Lambert solutions. Returns `None` if it
+    fails doing so.
+    """
+    try:
+        poly = lhs.as_poly()
+        gens = _filtered_gens(poly, symbol)
+        return _solve_lambert(lhs - rhs, symbol, gens)
+    except NotImplementedError:
+        pass
+
+
+def _solve_as_lambert(lhs, rhs, symbol, domain):
+    r"""
+    Helper solver to handle equations having Lambert solutions.
+    First tries to solve equation directly. If unsuccessful
+    attempts to find the solutions by making the `symbol` positive.
+    """
+    result = ConditionSet(symbol, Eq(lhs - rhs, 0), domain)
+
+    soln = _compute_lambert_solutions(lhs, rhs, symbol)
+    if soln is None:
+        # try with positive `symbol`
+        u = Dummy('u', positive=True)
+        pos_lhs = lhs.subs({symbol: u})
+        soln = _compute_lambert_solutions(pos_lhs, rhs, u)
```

Comment:
I don't think that this making the symbol positive trick should be used.

Reply:
> I don't think that this making the symbol positive trick should be used.

This code is from #14972. And it works for some equations and I'll try to come out with the reason.

Reply:
IMHO the most probable reason is the `_solve_lambert` give different outputs for real and positive symbols ... and there are a lot of conditions ( in order to make the suitable format of log terms else returning empty list (see from L-128 in `bivariate.py`)) in `_lambert(eq, x)` to make working more efficiently avoiding unnecessary solutions...

Reply:
If the equation should have negative solutions then this might lead to an incorrect result.

Reply:
> If the equation should have negative solutions then this might lead to an incorrect result.

Nice point!

Actually, this part is mostly giving the input to the log (e.g. `log(x)` --> `log(_u)` ) so I don't this so that negative solutions will be getting affected. In fact, it gives negative solutions even after passing through that logic.

```
input:
# this converts into log form 
solveset_real(x**3 - 3**x, x)

output:
# intermediate value : _R*log(3) - log(-_R**3) + I*pi 
# intermediate value (after passing through logic): _u*log(3) - 3*log(_u)
FiniteSet(-3*LambertW(-log(3)/3)/log(3), -3*LambertW(-log(3)/3, -1)/log(3))
# yes one more solution is missing i.e. 3 ( which will be after completion of other commits/PR's

# if the conditions are removed from `def _lambert(eq, x):` then the output will be ( here in this case no need for
# positive dummy (_u)
output:
FiniteSet((-LambertW(-log(3)))**(1/3)/log(3)**(1/3), -(-LambertW(-log(3)))**(1/3)/(2*log(3)**(1/3)) + sqrt(3)*I*(-LambertW(-log(3)))**(1/3)/(2*log(3)**(1/3)), -sqrt(3)*I*(-LambertW(-log(3)))**(1/3)/(2*log(3)**(1/3)) - (-LambertW(-log(3)))**(1/3)/(2*log(3)**(1/3)))

```
IMO, there should be another check with `if` the statement i.e. should be `real` and should have `log` to pass through this part.

Reply:
Please give your opinion on this.

Reply:
I think that whatever this substitution is doing should be achieved in a different way.

Reply:
> I think that whatever this substitution is doing should be achieved in a different way.

Any suggestions in general ?

Reply:
What about specifying that symbol is positive in the output of that answer (through the `Intersection` set) ?

Reply:
I think since most of the `_solve_lambert(f, symbol, gens)` function is related to solving the logarithms then passing a positive dummy should not be problem...
Also, I've noticed an irregularity that 

```
# the expansion of `log` term has different behaviour with 2 different dummies...
FOR 
solveset(x**2 - 2**x, x,Reals)
intermediate values

-_R*log(2) + log(-_R**2) - I*pi

-_u*log(2) + 2*log(_u) + 2*I*pi

& FOR
solveset(x*log(x) + 3*x + 1, x,Reals)
intermediate values
_R*(log(_R) + 3)
log(_R*log(_R) + 3*_R) - I*pi

_u*(log(_u) + 3)
log(_u) + log(log(_u) + 3) - I*pi

```

Reply:
The logs expand differently because the expansion is not always valid when the factors in the log are not positive:
```python
In [9]: x = Symbol('x', positive=True)

In [10]: f = log(x**2)

In [11]: expand_log(f).subs(x, -1)
Out[11]: 2⋅ⅈ⋅π

In [12]: f.subs(x, -1)
Out[12]: 0
```
If we expand the log when the assumptions are not satisfied then it can lead to a wrong result. Putting in a symbol with positive=True could lead to incorrect results being returned.

Reply:
https://github.com/sympy/sympy/pull/21025#discussion_r588132583

What about passing dummy just before log ( for sake of expansion) and then replacing back the symbol?!

Reply:
I think we should solve this issue later by removing this stuff and keep this under failed tests !